### ISSUE TITLE
Persistent congestion is not implemented

### ISSUE THREAD

**Issue by LiD0209** [cloudflare/quiche#2680]
# Persistent congestion is not implemented

## Summary
quiche detects ordinary packet/time-threshold loss, but does not implement the persistent congestion declaration required by [RFC 9002 Section 7.6](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6). When a valid persistent-congestion episode is reproduced, Reno, CUBIC, and BBR2/gcongestion all keep cwnd above `kMinimumWindow`.

## Standard Requirement

[RFC 9002 Section 7.6](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6), [Section 7.6.1](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6.1), and [Section 7.6.2](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6.2):

```text
When a sender establishes loss of all packets sent over a long enough
duration, the network is considered to be experiencing persistent
congestion.

The persistent congestion duration is computed as follows:

(smoothed_rtt + max(4*rttvar, kGranularity) + max_ack_delay) *
    kPersistentCongestionThreshold

A sender establishes persistent congestion after the receipt of an
acknowledgment if two packets that are ack-eliciting are declared
lost, and:

* across all packet number spaces, none of the packets sent between
  the send times of these two packets are acknowledged;
* the duration between the send times of these two packets exceeds
  the persistent congestion duration; and
* a prior RTT sample existed when these two packets were sent.

When persistent congestion is declared, the sender's congestion
window MUST be reduced to the minimum congestion window
(kMinimumWindow).
```

[RFC 9002 Appendix B.8](https://www.rfc-editor.org/rfc/rfc9002.html#appendix-B.8) also models this as filtering lost packets sent after `first_rtt_sample`, then setting:

```text
if (InPersistentCongestion(pc_lost)):
  congestion_window = kMinimumWindow
  congestion_recovery_start_time = 0
```

## Relevant Source Code

`quiche/src/recovery/congestion/recovery.rs:531` computes ordinary loss and forwards it directly to the selected congestion controller:

```rust
let loss = self.epochs[epoch].detect_lost_packets(
    loss_delay,
    self.pkt_thresh,
    now,
    trace_id,
    epoch,
);

if let Some(pkt) = loss.largest_lost_pkt {
    (self.congestion.cc_ops.congestion_event)(
        &mut self.congestion,
        self.bytes_in_flight.get(),
        loss.lost_bytes,
        &pkt,
        now,
    );
}
```

`quiche/src/recovery/gcongestion/recovery.rs:568` similarly removes lost packets and subtracts bytes, without a persistent-congestion check:

```rust
let loss_delay =
    self.rtt_stats.loss_delay(self.loss_thresh.time_thresh());

let LossDetectionResult {
    lost_bytes,
    lost_packets,
    pmtud_lost_bytes,
    pmtud_lost_packets,
} = self.epochs[epoch].detect_and_remove_lost_packets(
    loss_delay,
    self.loss_thresh.pkt_thresh(),
    now,
    lost,
);

self.bytes_in_flight
    .saturating_subtract(lost_bytes + pmtud_lost_bytes, now);
```

`quiche/src/recovery/congestion/cubic.rs:355` shows ordinary CUBIC loss response: it enters recovery and applies beta reduction, not mandatory collapse to `kMinimumWindow`:

```rust
if !in_congestion_recovery {
    r.congestion_recovery_start_time = Some(now);

    let ssthresh = (r.congestion_window as f64 * BETA_CUBIC) as usize;
    let ssthresh =
        cmp::max(ssthresh, r.max_datagram_size * MINIMUM_WINDOW_PACKETS);
    r.ssthresh.update(ssthresh, r.hystart.in_css());
    r.congestion_window = ssthresh;
}
```

`quiche/src/recovery/rtt.rs:35` tracks only `has_first_rtt_sample`; no stored `first_rtt_sample` timestamp is available for RFC B.8's `lost.time_sent > first_rtt_sample` filter:

```rust
pub struct RttStats {
    pub(super) latest_rtt: Duration,
    pub(super) smoothed_rtt: Duration,
    pub(super) rttvar: Duration,
    pub(super) min_rtt: Minmax<Duration>,
    pub(super) max_ack_delay: Duration,
    pub(super) has_first_rtt_sample: bool,
}
```

## Implementation Behavior

The recovery path can declare packets lost, count lost bytes, and run ordinary congestion events. It does not:

- compute `persistent_congestion_duration`;
- select two ack-eliciting lost packets as the bounding pair;
- check that no packets between their send times were acknowledged;
- require that the bounding packets were sent after the first RTT sample;
- set cwnd to `kMinimumWindow` or reset `congestion_recovery_start_time` when persistent congestion is established.

## Inconsistency Reason

[RFC 9002 Section 7.6.2](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6.2) requires an ACK-driven persistent-congestion check after lost packets are established. quiche stops at normal loss recovery, so a loss period that satisfies that section only causes Reno/CUBIC/BBR2 normal cwnd reduction. This fails the mandatory `congestion_window = kMinimumWindow` behavior.

## Runtime Evidence

The focused persistent-congestion test was run from `implementions/quiche`:

```text
cargo test -p quiche codex_tmp_persistent_congestion_collapses_cwnd --features gcongestion --release -- --nocapture
```

The command exited with code `101`, which is the expected reproducer signal for the failing assertions below.

The test first ACKs packet 0 to establish an RTT sample, then sends packets 1 through 9. ACKing packet 9 at 950 ms declares packets 1 through 8 lost; the 700 ms span between lost packets exceeds the persistent-congestion duration, so [RFC 9002 Section 7.6.2](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6.2) expects cwnd to become `2 * max_datagram_size = 2400`.

Observed output:

```text
bbr2_gcongestion: lost_packets=8 lost_bytes=8000 cwnd=14000 minimum_window=2400
cubic: lost_packets=8 lost_bytes=8000 cwnd=8400 minimum_window=2400
reno: lost_packets=8 lost_bytes=8000 cwnd=6000 minimum_window=2400

test result: FAILED. 0 passed; 3 failed
```

The failure is the expected reproducer signal: all three implementations report the required loss episode, but none collapses cwnd to `kMinimumWindow`.

## Impact

During sustained loss, quiche can continue with a cwnd larger than [RFC 9002 Section 7.6.2](https://www.rfc-editor.org/rfc/rfc9002.html#section-7.6.2) permits after persistent congestion. This can make the sender more aggressive than required under severe congestion.

## Fix Direction

Store the time of the first RTT sample. After ACK-driven loss detection, evaluate persistent congestion over lost ack-eliciting packets sent after that time, using the RFC duration formula and no-ACK-between-boundaries condition. If true, set cwnd to `kMinimumWindow` and reset congestion recovery start state.
